### PR TITLE
Jack/podcastdetail

### LINKED ISSUE #446: Refine PodcastDetailViewController
**Description:**
Make prettier (ask for more high fidelity designs) and prepare for future features (subscribing, sorting, separating unplayed)

### PR DESCRIPTION
**Pull request by jackthmp**
Closes #446 

- Redid `PodcastDetailViewController` with designs from Zeplin
- Connected search to home (via + button)
- Low level episode sorting (based on dates)

TODOs:
- ~Sticky header in `PodcastDetailViewController`~
- Loading indicators
- ~Date-string formatting~
- Connect to Drew's player

### CODE REVIEW

File: EpisodeFilterView.swift [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+import UIKit
+
+enum FilterType: String, CaseIterable {
+    case newest = "Newest"
+    case oldest = "Oldest"
+    case popular = "Popular"
+    case unlistened = "Unlistened"
+
+    func tag() -> (Int) {
```
Comment (mindylou) [inline EpisodeFilterView.swift]:
```suggestion
    func tag() -> Int {
```

File: EpisodeUtilityView.swift [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        playButton.titleLabel?.font = .systemFont(ofSize: 14)
+        playButton.setTitleColor(.gray, for: .normal)
+        playButton.setImage(#imageLiteral(resourceName: "play_icon"), for: .normal)
+
+        downloadButton = UIButton()
+        downloadButton.titleLabel?.font = .systemFont(ofSize: 14)
+        downloadButton.setTitleColor(.gray, for: .normal)
+
+        if isDownloaded {
+            downloadButton.setTitle("Downloaded", for: .normal)
```
Comment (mindylou) [inline EpisodeUtilityView.swift]:
```suggestion
            downloadButton.setTitle(isDownloaded ? "Downloaded" : "Download", for: .normal)
```
You can combine this if/else into two lines

File: EpisodeUtilityView.swift [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    func setUpConstraints() {
+        // MARK: - Constants
+        let sidePadding = 18
+        let playButtonSize = CGSize(width: 16, height: 18)
+        let downloadButtonSize = CGSize(width: 16, height: 16)
+
+        playButton.snp.makeConstraints { make in
+//            make.height.equalTo(buttonHeight)
```
Comment (mindylou) [inline EpisodeUtilityView.swift]: delete if you don't need it

File: PodcastDetailHeaderView.swift [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    var podcastGenres: [String]!
+
+    var filterView: EpisodeFilterView!
+
+    // MARK: - Constants
+    let cellReuseId = PodcastTagsCollectionViewCell.cellReuseId
+    let collectionViewEstimatedItemSize = CGSize(width: 64, height: 34)
+    let collectionViewItemSize = UICollectionViewFlowLayout.automaticSize
+    let collectionViewSectionInset = UIEdgeInsets(top: 0, left: 22, bottom: 0, right: 0)
+    let collectionViewMinimumInteritemSpacing = CGFloat(18)
```
Comment (mindylou) [inline PodcastDetailHeaderView.swift]:
```suggestion
    let collectionViewMinimumInteritemSpacing: CGFloat = 18
```

File: PodcastDetailHeaderView.swift [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+   func setUpConstraints() {
+        // MARK: - Constants
+        let imageHeight = 79
+        let imageTopPadding = UIApplication.shared.statusBarFrame.height + 25
+        let edgePadding = 36
+        let imageBottomPadding = 16.5
+        let titleHeight = 30
+        let publisherHeight = 21
+        let subscribeButtonPadding = 18
+        let subscribeButtonSize: CGSize = .init(width: 146, height: 34)
```
Comment (mindylou) [inline PodcastDetailHeaderView.swift]:
```suggestion
        let subscribeButtonSize = CGSize(width: 146, height: 34)
```

File: PodcastDetailHeaderView.swift [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        imageView.snp.makeConstraints { make in
+            make.top.equalToSuperview().offset(imageTopPadding)
+            make.height.width.equalTo(imageHeight)
+            make.centerX.equalToSuperview()
+        }
+
+        titleLabel.snp.makeConstraints { make in
+            make.top.equalTo(imageView.snp.bottom).offset(imageBottomPadding)
+            make.top.greaterThanOrEqualToSuperview().offset(imageTopPadding)
+
```
Comment (mindylou) [inline PodcastDetailHeaderView.swift]: remove blank line

File: Recast/Discover/MainSearchDataSourceDelegate.swift [hunk 7]
```diff
@@ -22,6 +22,7 @@ class MainSearchDataSourceDelegate: NSObject {
         SearchEndpoint(parameters: ["term": query, "media": "podcast", "limit": -1]).run()
             .success { response in
                 self.searchResults = response.results
+                print(response.results)
```
Comment (mindylou) [inline Recast/Discover/MainSearchDataSourceDelegate.swift]: remove print

File: Recast/PodcastDetailViewController.swift [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+    func filterEpisodes(filterType: FilterType) {
+        switch filterType {
+        case .newest: podcast?.items.sort { (e1, e2) -> Bool in
+            return e1.pubDate?.compare(e2.pubDate ?? Date()) == .orderedDescending
+            }
+        case .oldest: podcast?.items.sort { (e1, e2) -> Bool in
+            return e1.pubDate?.compare(e2.pubDate ?? Date()) == .orderedAscending
+            }
+        case .popular: break
```
Comment (mindylou) [inline Recast/PodcastDetailViewController.swift]:
```suggestion
        case .popular, .unlistened: break
```
combine the two cases into one

File: Recast/PodcastDetailViewController.swift [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+
+    override func didReceiveMemoryWarning() {
+        super.didReceiveMemoryWarning()
+        // Dispose of any resources that can be recreated.
+    }
+
+    func filterEpisodes(filterType: FilterType) {
+        switch filterType {
+        case .newest: podcast?.items.sort { (e1, e2) -> Bool in
+            return e1.pubDate?.compare(e2.pubDate ?? Date()) == .orderedDescending
```
Comment (mindylou) [inline Recast/PodcastDetailViewController.swift:149]: ooooo very nice

File: Recast/PodcastDetailViewController.swift [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+        super.viewDidLayoutSubviews()
+        gradientLayer.frame = backgroundImage.bounds
+    }
+
+    override func didReceiveMemoryWarning() {
+        super.didReceiveMemoryWarning()
+        // Dispose of any resources that can be recreated.
+    }
+
+    func filterEpisodes(filterType: FilterType) {
```
Comment (mindylou) [inline Recast/PodcastDetailViewController.swift]:
```suggestion
    func filterEpisodes(by filterType: FilterType) {
```
to make it more swifty

File: Recast/iTunesPodcasts/iTunesSearch/Models.swift [hunk 11]
```diff
@@ -35,7 +35,7 @@ struct SearchResult: Codable, PartialPodcast {
     let wrapperType: String
     let kind: String
     let collectionId: Int!
-    let trackId: Int
+//    let trackId: Int
```
Comment (mindylou) [inline Recast/iTunesPodcasts/iTunesSearch/Models.swift:38]: delete if unnecessary

**Review by mindylou — DISMISSED** [review]
looks good so far!

File: EpisodeFilterView.swift [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+        }
+
+        divider.snp.makeConstraints { make in
+            make.height.equalTo(dividerHeight)
+            make.leading.trailing.equalToSuperview()
+            make.centerY.equalTo(underline.snp.bottom)
+        }
+    }
+
+    @objc func didSelect(sender:UIButton) {
```
Comment (mindylou) [inline EpisodeFilterView.swift]:
```suggestion
    @objc func didSelect(sender: UIButton) {
```

File: EpisodeFilterView.swift [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+        selected = FilterType.allCases[sender.tag]
+        delegate?.filterEpisodes(by: selected)
+        UIView.animate(withDuration: 0.25) {
+            self.underline.snp.remakeConstraints { make in
+                make.height.equalTo(self.underlineHeight)
+                make.leading.trailing.equalTo(self.stackView.subviews[self.selected.tag()])
```
Comment (mindylou) [inline EpisodeFilterView.swift:105]: wow did not see this before... this is cool 💯

File: PodcastDetailHeaderView.swift [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
```
Comment (mindylou) [inline PodcastDetailHeaderView.swift]:
```suggestion
```

File: PodcastDetailHeaderView.swift [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+        }
+
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+}
+
```
Comment (mindylou) [inline PodcastDetailHeaderView.swift:140]: add `// MARK: - Collection view delegate / data source` to these extensions

File: Recast/PodcastDetailViewController.swift [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+//
+
+import UIKit
+import SnapKit
+import Kingfisher
+
+class PodcastDetailViewController: UIViewController, EpisodeFilterDelegate {
+
+    // MARK: - Variables
+    var partialPodcast: PartialPodcast!
```
Comment (mindylou) [inline Recast/PodcastDetailViewController.swift:16]: do we want to make an init with a `PartialPodcast` input? bc we wouldn't have a podcast detail view controller without a partial podcast

File: Recast/PodcastDetailViewController.swift [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+
+    override func viewDidAppear(_ animated: Bool) {
+        navigationController?.navigationBar.backgroundColor = .clear
+        navigationController?.navigationBar.setBackgroundImage(UIImage(), for: .default)
+        navigationController?.navigationBar.shadowImage = UIImage()
+        navigationController?.navigationBar.isTranslucent = true
+        Podcast.loadFull(from: self.partialPodcast, success: { podcast in
+            self.podcast = podcast
+            self.episodeTableView.reloadData()
+        }) { error in
```
Comment (mindylou) [inline Recast/PodcastDetailViewController.swift:99]: does this cause swiftlint errors?

File: Recast/iTunesPodcasts/iTunesSearch/Models.swift (hunk 11 above)
Comment (mindylou) [reply Recast/iTunesPodcasts/iTunesSearch/Models.swift:38]: +1

**Review by mindylou — APPROVED** [review]

**Review by jaewonsim — APPROVED** [review]
LGTM!

**Pull request merged**